core/vm/evm.go

```diff
@@ -70,18 +70,25 @@ func run(evm *EVM, contract *Contract, input []byte, readOnly bool) ([]byte, err
 // Context provides the EVM with auxiliary information. Once provided
 // it shouldn't be modified.
 type Context struct {
+	BlockContext *BlockContext
+	// Message information
+	Origin   common.Address // Provides information for ORIGIN
+	GasPrice *big.Int       // Provides information for GASPRICE
+}
+
+func (c *Context) GetHash(n uint64) common.Hash{
+	return c.BlockContext.GetHash(n)
+}
+
+// BlockContext provides the EVM with auxiliary
+// info about things that can be reused across transactions in a block
+type BlockContext struct {
```

Comment:
If the goal is just to cache canonical hashes, could you help me understand how this refactoring helps? It seems like all the good stuff is happening and can be contained with `GetHashFn`.

Reply:
So, the vm currently executes with a `vm.Context`. That context contains all the transaction specific data (msg.origin, gasprice) and also block-specific data. For every transaction, we create a new such object, thus throwing away the cache that was held in the getHash-closure. 

So this is a way to reuse the same base data across all transactions in a block

Reply:
Makes sense if you decide not to go the `BlockChain` encapsulated approach raised earlier. Thanks for clarifying.